### PR TITLE
Verify if configuration has changed before restarting Synapse

### PR DESCRIPTION
**Pull request by amandahla**
### Overview

Currently, most events trigger a restart even if the configuration has not changed. This will prevent that by checking the configuration before the restart.

### Rationale

Prevent constant restarts like we saw in Production handling database relation changed in second unit.

### Juju Events Changes

Config change event and integration related. Synapse will only restart if the generated configuration is different from the existing one in the workload.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Manging Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [ ] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

No need to update CH doc.

### CODE REVIEW

File: src/pebble.py [hunk 1]
```diff
@@ -276,6 +277,7 @@ def change_config(  # noqa: C901 pylint: disable=too-many-branches,too-many-stat
     """
     try:
         synapse.execute_migrate_config(container=container, charm_state=charm_state)
+        existing_synapse_config = _get_synapse_config(container)
         current_synapse_config = _get_synapse_config(container)
```
Comment (javierdelapuente) [inline src/pebble.py:281]: isn't `existing_synapse_config` equal to `current_synapse_config`?, they are just one line apart

**Review by javierdelapuente — COMMENTED** [review]
Maybe a unit test could be added for the change?

File: src/pebble.py (hunk 1 above)
Comment (mthaddon) [inline src/pebble.py:281]: I'm not sure I understand how these two values would ever be different - they're running the same things?

Comment (mthaddon) [reply src/pebble.py:281]: Ok, I see how this works, you're passing current_synapse_config to enable_* methods below, which may modify it.

File: src/pebble.py [hunk 2]
```diff
@@ -324,16 +326,27 @@ def change_config(  # noqa: C901 pylint: disable=too-many-branches,too-many-stat
...
+                container,
+                get_worker_config(unit_number),
+                config_path=synapse.SYNAPSE_WORKER_CONFIG_PATH,
+            )
+            # Push main configuration
+            _push_synapse_config(container, current_synapse_config)
+            synapse.validate_config(container=container)
+            restart_synapse(container=container, charm_state=charm_state, is_main=is_main)
+        else:
+            logging.info("Configuration has not changed, no action.")
```
Comment (mthaddon) [inline src/pebble.py:349]: I think it'd be good to have a test that confirms this behaviour. Confirm that if a config-changed hook fires but configuration doesn't change the service isn't restarted, and vice versa.

**Review by javierdelapuente — APPROVED** [review]
LGTM! tests for the next PR

File: src/pebble.py (hunk 2 above)
Comment (amandahla) [reply src/pebble.py:349]: Nice, I'll add one in further PR.

**Comment by amandahla** [comment]
I'll add a unit test in further PR.

**Pull request merged**